File: liveq/lib/utils/services.dart
import 'dart:async';

import 'package:audioplayers/audioplayers.dart';
import 'package:flutter/material.dart';
import 'package:flutter/foundation.dart' show kIsWeb;
import 'package:http/http.dart' as http;
import 'package:liveq/models/player.dart';
import 'package:liveq/models/catalog.dart';
import 'dart:convert' as convert;
import 'package:liveq/utils/utils.dart';

import 'package:spotify/spotify.dart';
import 'package:spotify_sdk/spotify_sdk.dart';

import 'package:liveq/utils/song.dart';

abstract class Service {
  static const String SOUNDCLOUD = 'SoundCloud';
  static const String SPOTIFY = 'Spotify';
  static const String APPLE = 'Apple Music';

  String name;
  bool isConnected = false;
  // Future<bool> get isConnected;

  static final List<Service> potentialServices = [Spotify(), SoundCloud()];

  String iconImagePath;

  // void Function(void) callback;

  Future<bool> connect();
  // Future<bool> connect() async {
  //   connectedServices.add(this);
  //   await saveService();
  //   isConnected = true;
  //   return Future.value(true);
  // }

  Future<void> play(String uri, PlayerModel player);
  Future<void> resume();
  Future<void> pause();

  Widget getImageIcon() {
    if (iconImagePath != null) {
      return ImageIcon(AssetImage(iconImagePath));
    } else {
      return Icon(Icons.music_note);
    }
  }

  Future<List<Song>> search(String query);

  static Service fromString(String s) {
    Service service;

    switch (s) {
      case SPOTIFY:
        service = Spotify();
        break;
      case SOUNDCLOUD:
        service = SoundCloud();
        break;
      case APPLE:
        break;
    }

    return service;
  }
}

class SoundCloud extends Service {
  final String name = CatalogModel.SOUNDCLOUD;

  // final String clientId = 'YaH7Grw1UnbXCTTm0qDAq5TZzzeGrjXM';
  final String clientId = 'e38841b15b2059a39f261df195dfb430';
  final String userId = '857371-474509-874152-946359';
  final String iconImagePath = 'assets/images/soundcloud.png';

  static final Service _soundcloud = SoundCloud._internal();

  // AudioStream stream;
  AudioPlayer player = AudioPlayer();

  SoundCloud._internal();

  factory SoundCloud() {
    return _soundcloud;
  }

  @override
  Future<bool> connect() async {
    // super.connect();
    // TODO: implement connect
    // throw UnimplementedError();
    return Future.value(true);
  }

  @override
  Future<void> pause() async {
    // TODO: implement pause
    // stream.pause();
    print("pausing");
    player.pause();
  }

  @override
  Future<void> play(String id, PlayerModel playerApi) async {
    String uri =
        'https://api.soundcloud.com/tracks/$id/stream?client_id=$clientId';
    print('playing $uri');
    AudioPlayer.logEnabled = true;
    int result = await player.play(uri);
    if (result == 1) {
      player.onPlayerCompletion.listen((event) {
        playerApi.next();
      });
    } else {
      print("FAIL");
    }
  }

  @override
  Future<void> resume() {
    // TODO: implement resume
    // stream.resume();
    player.resume();
  }

  @override
  Future<List<Song>> search(String query) async {
    
    List<Song> searchResults = List();
    String search =
        'https://api.soundcloud.com/tracks?q=${formatSearch(query)}&limit=100&format=json&client_id=$clientId';

    var response = await http.get(search);

    if (response.statusCode == 200) {
      var jsonResponse = convert.jsonDecode(response.body);

      for (var item in jsonResponse) {
        if (item['kind'] == 'track' && item['artwork_url'] != null) {
          var track = item;
          String trackId = track['id'].toString();
          String uri = track['uri'];
          String trackName = track['title'];
          String artist = track['user']['username'];
          String imageUri = track['artwork_url'];
          int duration = track['duration'];
          Service service = this;

          Song song = Song(0, trackId, uri, trackName, artist, imageUri, duration, service);
          searchResults.add(song);
        }
      }
    } else {
      print('ERROR');
      print(response.statusCode);
    }

    return searchResults;
  }

  String formatSearch(String query) {
    String s = query.replaceAll(' ', '%20');
    s = s.replaceAll('(', '%28');
    s = s.replaceAll(')', '%29');
    return s;
  }
}

class Spotify extends Service {
  final String name = CatalogModel.SPOTIFY;
  final String iconImagePath = 'assets/images/Spotify_Icon_RGB_Green.png';

  // Developer tokens. DO NOT CHANGE
  final String clientId = '03237b2409b24752a3f0c33262ad2d02';
  final String clientSecret = '52560cee72394fc5a049731f2d8f001e';
  final String redirectUri = 'spotify-ios-quick-start://spotify-login-callback';

  SpotifyApi spotifyWebApi;

  static final Service _spotify = Spotify._internal();

  Spotify._internal();

  factory Spotify() {
    return _spotify;
  }

  /// Connect to the SpotifySDK and get an [authenticationToken]
  @override
  Future<bool> connect() async {
    // super.connect();
    // Use the spotify package to create credentials. This is only needed for Search
    var credentials = SpotifyApiCredentials(clientId, clientSecret);
    spotifyWebApi = SpotifyApi(credentials);

    // Use the spotify_sdk package if on mobile to allow playing
    if (!kIsWeb && isHost) {
      await SpotifySdk.connectToSpotifyRemote(
          clientId: this.clientId, redirectUrl: this.redirectUri);
    }

    return spotifyWebApi != null;
  }

  /// Pause a [Song]
  @override
  Future<void> pause() async {
    await SpotifySdk.pause();
  }

  /// Play a [Song] given a [uri]
  @override
  Future<void> play(String uri, PlayerModel player) async {
    await SpotifySdk.play(spotifyUri: uri);
    SpotifySdk.subscribePlayerContext().listen((playerContext) {
      if (playerContext.uri.contains('station')) {
        player.next();
      }
    });
  }

  /// Resume the [Song] playing right now
  @override
  Future<void> resume() async {
    await SpotifySdk.resume();
  }

  /// Use the [authenticationToken] to search a query using the Spotify Web API
  @override
  Future<List<Song>> search(String query) async {
    List<Song> searchResults = List();

    var search =
        await spotifyWebApi.search.get(query).first().catchError((err) {
      print('SPOTIFY ERROR: ${(err as SpotifyException).message}');
    });

    if (search != null) {
      search.forEach((pages) {
        pages.items.forEach((track) async {
          if (track is Track) {
            String trackId = track.id;
            String uri = track.uri;
            String trackName = track.name;
            List<String> _artistNames =
                track.artists.map((val) => val.name).toList();
            String artists = _artistNames.join(", ");
            String imageUri = track.album.images[1].url;
            int duration = track.durationMs;
            Service service = this;

            searchResults.add(
                Song(0, trackId, uri, trackName, artists, imageUri, duration, service));
          }
        });
      });
    }

    return searchResults;
  }
}

// class Apple extends Service {
//   final String name = Service.APPLE;
//   final String iconImagePath = 'assets/images/Apple_Music_Icon.psd';

//   static final Service _apple = Apple._internal();

//   Apple._internal();

//   factory Apple() {
//     return _apple;
//   }

//   @override
//   Future<bool> connect() {
//     // super.connect();
//     // TODO: implement connect
//     throw UnimplementedError();
//   }

//   @override
//   Future<void> pause() {
//     // TODO: implement pause
//     throw UnimplementedError();
//   }

//   @override
//   Future<void> play(String uri) {
//     // TODO: implement play
//     throw UnimplementedError();
//   }

//   @override
//   Future<void> resume() {
//     // TODO: implement resume
//     throw UnimplementedError();
//   }

//   @override
//   Future<List<Song>> search(String uri) {
//     // TODO: implement search
//     throw UnimplementedError();
//   }
// }

File: ios/LiveQ/LiveQ/Utils/Services.swift
//
//  Services.swift
//  LiveQ
//
//  Created by Shaheer Mirza on 4/19/20.
//  Copyright © 2020 Shaheer Mirza. All rights reserved.
//

import Foundation
import AVKit
import Alamofire

enum ServiceType: String {
    case Spotify, SoundCloud, Apple
}

protocol Service {
    
    static var instance: Service { get }
    var name: String { get }
    var isSelected: Bool { get set }
    var isConnected: Bool { get set }
    var isAllowed: Bool { get set }
    var image: UIImage { get }
    
    func connect()
    func play(_ uri: String)
    func resume()
    func pause()
    func stop()
    func search(query: String, finished: @escaping (_ songs: [Song]) -> Void)
}

extension Service {
    func formatSearch(query: String) -> String {
        var s: String = query.replacingOccurrences(of: " ", with: "%20", options: .literal, range: nil)
        s = s.replacingOccurrences(of: "(", with: "%28", options: .literal, range: nil)
        return s.replacingOccurrences(of: ")", with: "%29", options: .literal, range: nil)
    }
}

class SoundCloud: Service {
    static let instance: Service = SoundCloud()
    private init() {}
    
    var isConnected: Bool = true
    var isAllowed: Bool = true
    
    var name: String = "SoundCloud"
    var isSelected: Bool = true
    var image: UIImage = UIImage(named: "soundcloud.png")!
    
    final var clientId: String = "e38841b15b2059a39f261df195dfb430";
    final var playId: String = "e38841b15b2059a39f261df195dfb430";
    final var userId: String = "857371-474509-874152-946359";
    
    var player: AVPlayer?
    
    func connect() {
        // do nothing
        self.isConnected = true
    }
    
    func play(_ id: String) {
        let playUri: String = "https://api.soundcloud.com/tracks/\(id)/stream?client_id=\(clientId)"
        print("wanting to play \(playUri)")
        player = AVPlayer(url: URL(string: playUri)!)
        player?.play()
        NotificationCenter.default.addObserver(Player.instance, selector: #selector(Player.instance.next), name: .AVPlayerItemDidPlayToEndTime, object: nil)
    }
    
    func resume() {
        player?.play()
    }
    
    func pause() {
        player?.pause()
    }
    
    func stop() {
        print("SC stopping")
        player?.replaceCurrentItem(with: nil)
        player = nil
    }
    
    func search(query: String, finished: @escaping (_ songs: [Song]) -> Void) {
        
        let search = "https://api.soundcloud.com/tracks?q=\(formatSearch(query: query))&limit=100&format=json&client_id=\(clientId)"
        
        var request: URLRequest = URLRequest(url: URL(string: search)!)
        request.httpMethod = "GET"
        
        let task = URLSession.shared.dataTask(with: request) { data, response, error in
            guard let data = data, error == nil else { return }
            
            var searchResults = [Song]()
            
            guard let json = try? JSONSerialization.jsonObject(with: data, options: JSONSerialization.ReadingOptions()) as? [AnyObject] else {
                print("ERROR: conversion from JSON failed")
                return
            }
            
            for item in json {
                guard let track = item as? [String: Any] else { return }
                let artworkUrl: String = track["artwork_url"] as? String ?? ""
                let kind: String = track["kind"] as? String ?? ""

                if !kind.isEmpty && !artworkUrl.isEmpty {
                    let artistJson = track["user"]
                    let artistConvert = artistJson as! [String: Any]
                    let artistObject = Artist.init(name: artistConvert["username"] as! String)
                    
                    let id = "\(track["id"] ?? 0)"
                    let uri = track["uri"] as! String
                    let trackName = track["title"] as! String
                    let artist = artistObject
                    let imageUri = artworkUrl
                    let duration = track["duration"] as! Int
                    let service = self
                    
                    let song = Song(
                        id: Int(id) ?? Int.random(in: 1..<1500),
                        trackId: id,
                        uri: uri,
                        name: trackName,
                        artists: [artist],
                        imageUri: imageUri,
                        duration: duration,
                        service: service
                    )
                    
                    searchResults.append(song)
                }
            }
            finished(searchResults)
        }
        task.resume()
    }
    
    func testTrack(id: String, finished: @escaping ((_ isSuccess: Bool) -> Void)) {
        let testUrl = "https://api.soundcloud.com/tracks/\(id)?client_id=\(playId)"
        var request: URLRequest = URLRequest(url: URL(string: testUrl)!)
        request.httpMethod = "GET"
        
        let task = URLSession.shared.dataTask(with: request) { data, response, error in
            if let response = response {
                let httpResponse = response as! HTTPURLResponse
                
                if httpResponse.statusCode == 200 {
                    finished(true)
                } else {
                    finished(false)
                }
            }
        }
        task.resume()
    }
}

class Spotify: NSObject, Service, SPTAppRemoteDelegate {
    
    var isSelected: Bool = false
    var isConnected: Bool = false
    var isAllowed: Bool = false
    
    var image: UIImage = UIImage(named: "spotify")!
    
    static let instance: Service = Spotify()
    
    private override init() {}
    
    var name: String = "Spotify"
    
    var playerState: SPTAppRemotePlayerState?
    var playerStateDelegate: PlayerStateDelegate?
    
    static private let kAccessTokenKey = "access-token-key"
    private let redirectUri = URL(string: "spotify-ios-quick-start://spotify-login-callback")!
    private let clientIdentifier = "03237b2409b24752a3f0c33262ad2d02"
    private let clientSecret = "52560cee72394fc5a049731f2d8f001e"
    private var trackIdentifier = "spotify:track:7p5bQJB4XsZJEEn6Tb7EaL"
    
    private var webToken: String?
    
    lazy var configuration: SPTConfiguration = {
        let configuration = SPTConfiguration(clientID: clientIdentifier, redirectURL: redirectUri)
        configuration.playURI = self.trackIdentifier
        configuration.tokenSwapURL = URL(string: "http://localhost:1234/swap")
        configuration.tokenRefreshURL = URL(string: "http://localhost:1234/refresh")
        return configuration
    }()
    
    lazy var appRemote: SPTAppRemote = {
        let appRemote = SPTAppRemote(configuration: self.configuration, logLevel: .debug)
        appRemote.connectionParameters.accessToken = self.accessToken
        appRemote.delegate = self
        return appRemote
    }()
    
    var accessToken = UserDefaults.standard.string(forKey: kAccessTokenKey) {
        didSet {
            let defaults = UserDefaults.standard
            defaults.set(accessToken, forKey: Spotify.kAccessTokenKey)
        }
    }
    
    func connect() {
        
//        let parameters = ["client_id": self.clientIdentifier,
//                          "client_secret": self.clientSecret,
//                          "grant_type": "client_credentials"]
//
//        AF.request("https://accounts.spotify.com/api/token", method: .post, parameters: parameters).responseJSON(completionHandler: { response in
//
//            print(response)
//            print(response.result)
//            let jsonData = response.value as! NSDictionary
//            let token = jsonData.value(forKey: "access_token") as? String
//            self.accessToken = token!
//        })
        
        if Player.instance.isHost {
            print("authorize and play")
//            self.appRemote.authorizeAndPlayURI(trackIdentifier)
//            self.isConnected = self.accessToken != nil
            print("here")
//            DispatchQueue.main.async {
            self.appRemote.authorizeAndPlayURI(self.trackIdentifier)
            print("connecting on \(Thread.current)")
            print(self.appRemote.connectionParameters.accessToken)
            self.appRemote.connect()
//            }
//            DispatchQueue.main.schedule(after: ) {
//                self.appRemote.connect()
//            }
//            if let _ = self.appRemote.connectionParameters.accessToken {
//                print("reconnecting")
//                self.appRemote.connect()
//                self.isConnected = self.accessToken != nil
//            } else {
//                print("authorize and play")
//                self.appRemote.authorizeAndPlayURI(trackIdentifier)
//                self.isConnected = self.accessToken != nil
//            }
            
            print("pausing")
            print(self.appRemote.playerAPI == nil)
            self.appRemote.playerAPI?.pause()
        }
    }
    
    func play(_ uri: String) {
        print("spotify playing \(uri)")
        print("connected: \(self.appRemote.isConnected)")
        self.trackIdentifier = uri
        if !self.appRemote.isConnected {
            self.appRemote.authorizeAndPlayURI(uri)
        } else {
            self.appRemote.playerAPI?.play(uri)
        }
        
        SceneDelegate.shared?.alreadyConnected = true
        
        
    }
    
    func resume() {
        self.appRemote.playerAPI?.resume()
    }
    
    func pause() {
        self.appRemote.playerAPI?.pause()
    }
    
    func stop() {
        
    }
    
    func search(query: String, finished: @escaping (_ songs: [Song]) -> Void) {
        var searchResults = [Song]()
        
        var search = "https://api.spotify.com/v1/search/?type=track&market=US&q="
        search += formatSearch(query: query)
        
        var request: URLRequest = URLRequest(url: URL(string: search)!)
        request.httpMethod = "GET"
        request.setValue("Bearer \(self.accessToken!)", forHTTPHeaderField: "Authorization")

        let task = URLSession.shared.dataTask(with: request) { (data, response, error) in
            if let error = error {
                print("Error while searching: \(error)")
                return
            }

            let parsedResult = try? JSONDecoder().decode(SearchResult.self, from: data!)
            if let results = parsedResult {
                searchResults = results.getSongs()
                print("found \(searchResults.count) results")
                finished(searchResults)
            }
        }
        task.resume()
    }
    
    func setToken(openURLContexts URLContexts: Set<UIOpenURLContext>) {
        
        guard let url = URLContexts.first?.url else { return }
        
        let parameters = appRemote.authorizationParameters(from: url)
        
        if let access_token = parameters?[SPTAppRemoteAccessTokenKey] {
            appRemote.connectionParameters.accessToken = access_token
            self.accessToken = access_token
        } else if let error_description = parameters?[SPTAppRemoteErrorDescriptionKey] {
            print(error_description)
        }
    }
    
    func appRemoteDidEstablishConnection(_ appRemote: SPTAppRemote) {
        self.appRemote.playerAPI?.delegate = Player.instance
        self.appRemote.playerAPI?.subscribe(toPlayerState: { (result, error) in
            if let error = error {
                print(error.localizedDescription)
            }
        })
        self.appRemote.playerAPI?.setRepeatMode(.off)
        print("established connection \(Thread.current)")
        
    }
    
    func appRemote(_ appRemote: SPTAppRemote, didFailConnectionAttemptWithError error: Error?) {
        self.appRemote.authorizeAndPlayURI(trackIdentifier)
        self.appRemote.connect()
        print("failed connected \(error)")
    }
    
    func appRemote(_ appRemote: SPTAppRemote, didDisconnectWithError error: Error?) {
        
        print("disconnectedWithError: \(error)")
    }
    
    func disconnect() {
        if self.appRemote.isConnected {
            self.appRemote.disconnect()
            self.isConnected = false
        }
    }
    
}
